### ISSUE TITLE
warmup_frac is correct but no config sets it, so the drift 7642867 describes is still live in every run

### ISSUE THREAD

**Issue by kuanlinhuang** [Huang-lab/TRACE-Kin#14]
`7642867` identifies a real problem and fixes it correctly. This is about the last step.

Also: thanks for the replies on #13 — the cache answer in particular explains the recompute in a way that makes it a choice rather than an oversight, which is what that point was asking.

## The commit is right, including the part that is easy to get wrong

`warmup_iters` is absolute while `lr_decay_iters` auto-scales to `total_iters`, so the warmup *fraction* moves whenever the budget does. The ordering the commit message claims does hold:

```python
if total_iters:
    self.total_iters = total_iters
else:
    self.total_iters = num_epochs * steps_per_epoch
...
self.warmup_iters = (int(warmup_frac * self.total_iters) if warmup_frac else warmup_iters)
```

That matters more than it looks. The only caller, `train_trace_kin.py:932`, passes `total_iters=None`, so the `else` branch is the live one — and it is the branch that carries the subsample scaling through `steps_per_epoch`. Had the two blocks been the other way round, every run with `warmup_frac` set would have died on `None * float`.

## 1. Nothing turns it on

```
$ grep -rn "warmup_frac" --include=*.json --include=*.yaml .
(no matches)

$ grep -rn "warmup_iters" --include=*.json training/
training/config_v6c.json:17:     "warmup_iters": 500,
training/config_v7.json:17:      "warmup_iters": 500,
training/config_v7_grit.json:20: "warmup_iters": 500,
training/config_v1.json:17:      "warmup_iters": 0,     # v4, v5, v5t likewise
```

The mechanism is in and unused, so every run today still takes the absolute count. A v7 ablation at 33% subsample still gets 3x the warmup fraction of the full-data run — exactly the comparison the commit message says is being corrupted.

I did not send a PR for this because the right value depends on `steps_per_epoch`, which depends on your dataset and batch size, and I cannot compute it from the repository. If 500 is the intended full-data warmup, then `"warmup_frac": 500 / total_iters` at the full-data setting leaves the full-data schedule byte-identical and moves only the subsampled arms — which is the point. Happy to send the three-line change if you tell me the number, or if you would rather it be computed at startup and logged.

Worth doing before the next sweep rather than after: a sweep run under the current configs cannot be compared to one run after the change.

## 2. `warmup_frac: 0` silently means "unset"

```python
self.warmup_iters = (int(warmup_frac * self.total_iters)
                     if warmup_frac else warmup_iters)
```

Truthiness, so `0` falls through to `warmup_iters` rather than giving zero warmup. Three lines below, the same constructor treats `0` as a deliberate value and says why:

```python
# Treat both None and 0 as "use total_iters" so a missing lr_decay_iters
# in config does not collapse the LR to min_lrate immediately after warmup.
if lr_decay_iters is None or lr_decay_iters == 0:
```

Not hypothetical: four configs in this repo already use `warmup_iters: 0` to mean no warmup, so `warmup_frac: 0` is the natural way to write the no-warmup arm of a sweep. It would silently give 500 steps instead, and the job log would not say so.

`if warmup_frac is not None` fixes it. While there, nothing clamps the value — `warmup_frac: 1.0` warms up for the whole run and the LR never decays, also silently.

## Still open: #12

Unrelated to this commit, but it has had no reply and the configs are unchanged:

```
$ grep -rn "mol_pe_raw_norm" --include=*.json training/
training/config_v7.json:55:      "mol_pe_raw_norm" : "batch"
training/config_v7_grit.json:71: "mol_pe_raw_norm": "batch",
```

The measurement in #12 is 48% of the ligand embedding's norm at the `config_v7_grit.json` operating point, against a reference that defaults the norm off and warns against it in source. The part that does not keep is `config_v7.json` reading as the baseline while being the widened+BatchNorm arm: a run recorded as `config_v7.json` before 2026-09-03 and one after are not the same model, and re-running cannot reveal it because the code is byte-identical.

If the BatchNorm arm is what you want, renaming the config says so in one move and makes every recorded run interpretable again. If it is not, `"none"` in `config_v7_grit.json` starts the GRIT arm from the published operating point. Either is fine; the current state is the one that is not, and it costs you the baseline rather than costing anyone time.

From the 2026-09-17 Huang Lab code review.
